snuba/utils/streams/kafka.py

```diff
@@ -45,35 +67,70 @@ class KafkaConsumer(Consumer[TopicPartition, int, bytes]):
+            self.__resolve_partition_starting_offset = self.__resolve_partition_offset_earliest
+        elif auto_offset_reset in {"largest", "latest", "end"}:
+            self.__resolve_partition_starting_offset = self.__resolve_partition_offset_latest
+        elif auto_offset_reset == "error":
+            self.__resolve_partition_starting_offset = self.__resolve_partition_offset_error
+        else:
+            raise ValueError("invalid value for 'auto.offset.reset' configuration")
+
+        self.__consumer = ConfluentConsumer(
+            {**configuration, "auto.offset.reset": "error"}
+        )
 
-    def __wrap_assignment_callback(
-        self, callback: Callable[[Sequence[TopicPartition]], None]
-    ) -> Callable[[ConfluentConsumer, Sequence[ConfluentTopicPartition]], None]:
-        def wrapper(
-            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
-        ) -> None:
-            callback(
-                [TopicPartition(value.topic, value.partition) for value in partitions]
-            )
+    def __resolve_partition_offset_earliest(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        low, high = self.__consumer.get_watermark_offsets(partition)
+        return ConfluentTopicPartition(partition.topic, partition.partition, low)
+
+    def __resolve_partition_offset_latest(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        low, high = self.__consumer.get_watermark_offsets(partition)
+        return ConfluentTopicPartition(partition.topic, partition.partition, high)
 
-        return wrapper
+    def __resolve_partition_offset_error(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        raise ConsumerError
 
     def subscribe(
         self,
         topics: Sequence[str],
         on_assign: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
         on_revoke: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
     ) -> None:
-        kwargs = {}
+        def assignment_callback(
+            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
+        ) -> None:
+            assignment: MutableSequence[ConfluentTopicPartition] = []
+
+            for partition in self.__consumer.committed(partitions):
+                if partition.offset >= 0:
+                    assignment.append(partition)
+                elif partition.offset == OFFSET_INVALID:
+                    assignment.append(self.__resolve_partition_starting_offset(partition))
+                else:
+                    raise ValueError("received unexpected offset")
+
+            self.__consumer.assign(assignment)
 
-        if on_assign is not None:
-            kwargs["on_assign"] = self.__wrap_assignment_callback(on_assign)
+            if on_assign is not None:
+                on_assign([TopicPartition(i.topic, i.partition) for i in partitions])
```

Comment:
I realized there are a couple flaws with this while I was working on test for the `seek`/`tell` change.

The exceptions that are raised here is catchable via a `try`/`except` around `KafkaConsumer.poll`. If the exception doesn't propagate in a way that it exits the scope where the consumer is defined, there is nothing preventing `poll` from being called again. (As far as I can tell, throwing from this callback doesn't inherently throw the consumer into an unrecoverable state.)

I'm a little concerned about the potential for users to naively discard `ConsumerError`.  If this happens:

- `assign` will never have been called, so offsets will not have been reset to the committed offset (for *any* partition, since that occurs outside of the loop body), or using the starting offset resolution strategy
- `on_assign` will never have been called

Reply:
> (As far as I can tell, throwing from this callback doesn't inherently throw the consumer into an unrecoverable state.)

Just confirmed that throwing in the callback does not prevent future `poll` calls from returning results.

Reply:
Why not try catching inside this method, setting a private attribute on this class that would make a subsequent call to poll fail? It feels a little complex, but if this method needs to successfully complete to keep the Consumer state in a clean state, and the Confluent consumer does not enforce this for us, we have to protect the Consumer state ourselves.

Reply:
This does seem like it's probably the safest option, though it does mean that a single bad partition will spoil the whole bunch, so to speak. I'm not a big fan of that but I also don't really have any other ideas.

Reply:
so, to be on the same page, are you planning to make a change here ?

Reply:
Yes, my plan is to follow through with what you suggested above, but am still open to alternative options for the longer term solution.

Reply:
I brought over the consumer state concept from #526 which seems like a good fit here.

I can't decide if this should close the consumer as part of this `except` block as well to make sure the assignment is released here. Thoughts?

Reply:
Cannot find a strong reason one way or the other. Even if we do not close, we are in an unfixable error condition that is likely to bring the consumer process to halt one way or another, so assignments are going to be released.

Reply:
Ok, I'll leave it as-is for now.